librad/src/git/local/transport.rs

```diff
@@ -0,0 +1,388 @@
+    },
+    keys::SecretKey,
+    paths::Paths,
+    uri::RadUrn,
+};
+
+lazy_static! {
+    static ref SETTINGS: Arc<RwLock<Option<Settings>>> = Arc::new(RwLock::new(None));
+}
+
+#[derive(Clone)]
+pub struct Settings {
+    pub paths: Paths,
+    pub signer: SecretKey,
+}
+
+pub fn register(settings: Settings) {
+    static INIT: Once = Once::new();
+
+    LocalTransportFactory::new().configure(settings);
+    unsafe {
+        INIT.call_once(move || {
+            git2::transport::register(local::URL_SCHEME, move |remote| {
+                Transport::smart(&remote, true, LocalTransportFactory::new())
+            })
+            .unwrap()
+        });
+    }
+}
+
+#[derive(Clone)]
+struct LocalTransportFactory {
+    settings: Arc<RwLock<Option<Settings>>>,
+}
+
+impl LocalTransportFactory {
+    fn new() -> Self {
+        Self {
+            settings: SETTINGS.clone(),
+        }
+    }
+
+    fn configure(&self, settings: Settings) {
+        *self.settings.write().unwrap() = Some(settings)
+    }
+}
+
+impl SmartSubtransport for LocalTransportFactory {
+    fn action(
+        &self,
+        url: &str,
+        service: Service,
+    ) -> Result<Box<dyn SmartSubtransportStream>, git2::Error> {
+        let settings = self.settings.read().unwrap();
+        match *settings {
+            None => Err(git2::Error::from_str("local transport unconfigured")),
+            Some(ref settings) => {
+                let url = url.parse::<LocalUrl>().map_err(into_git_err)?;
+
+                let mut transport = LocalTransport::new(settings.clone()).map_err(into_git_err)?;
+                let stream = transport
+                    .stream(url, service, Localio::piped())
+                    .map_err(into_git_err)?;
+
+                Ok(Box::new(stream))
+            },
+        }
+    }
+
+    fn close(&self) -> Result<(), git2::Error> {
+        Ok(())
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Mode {
+    Stateless,
+    Stateful,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("No such URN: {0}")]
+    NoSuchUrn(RadUrn),
+
+    #[error(transparent)]
+    Storage(#[from] storage::Error),
+
+    #[error(transparent)]
+    Git(#[from] git2::Error),
+
+    #[error(transparent)]
+    Io(#[from] io::Error),
+}
+
+/// A running service (as per the [`Service`] argument) with it's stdio
+/// connected as per [`Localio`].
+///
+/// [`Connected::wait`] MUST be called, in order to `wait(2)` on the child
+/// process, and run post-service hooks.
+#[must_use = "`wait` must be called"]
+pub struct Connected {
+    process: Child,
+    on_success: Option<Box<dyn FnOnce() -> Result<(), Error> + Send + 'static>>,
+}
+
+impl Connected {
+    pub fn wait(&mut self) -> Result<(), Error> {
+        let status = self.process.wait()?;
+        if status.success() {
+            match self.on_success.take() {
+                None => Ok(()),
+                Some(f) => f(),
+            }
+        } else {
+            Ok(())
```

Comment:
Why `Ok(())` if the child process failed?

Reply:
605ec4cba2d2866ae4e8c026bc79e795e42c3cfc